### ISSUE TITLE
Allow Header(limit=None) ?

### ISSUE THREAD

**Issue by NicolasHug** [pytorch/data#762]
Not urgent at all, just a minor suggestion:

In the benchmark scripts I'm currently running I want to limit the number of samples in a data-pipe according to an `args.limit` CLI parameter. I'd be nice to be able to just write:

```py
dp = Header(dp, limit=args.limit)
```

and let `Header` be a no-op when `limit=None`. This might be a bit niche, and the alternative is to just protect the call in a `if` block, so I would totally understand if this isn't in scope (and it's really not urgent in any case)

**Comment by mabu-github** [comment]
I had a quick look at it and during the implementation noticed the following: it is probably more appropriate to leave it up to the caller to determine whether to limit the elements (using `Header`) or not (using the source datapipe without wrapping it into `Header`).

Using Header is an operation that needs a limit. Otherwise the `__len__` function is undefined for the case of an unbounded source datapipe when the limit would be set to `None`.

See `torchdata/datapipes/iter/util/header.py`:
```
def __len__(self) -> int:
    if self.length != -1:
        return self.length
    try:
        source_len = len(self.source_datapipe)
        if self.limit is None:
            self.length = source_len
        else:
            self.length = min(source_len, self.limit)
        return self.length
    except TypeError:
        warn(
            "The length of this HeaderIterDataPipe is inferred to be equal to its limit."
            "The actual value may be smaller if the actual length of source_datapipe is smaller than the limit."
        )
        # (!) limit could be None at this time which breaks the contract of __len__
        # (!) there is no way to determine a length because the source datapipe could be endless
        return self.limit
```

Therefore we should think about whether to move on with this or not because the feature seems to break the software design.

**Comment by ejguan** [comment]
> I had a quick look at it and during the implementation noticed the following: it is probably more appropriate to leave it up to the caller to determine whether to limit the elements (using `Header`) or not (using the source datapipe without wrapping it into `Header`).

This is true for the case that the whole pipeline is created by users. There can be a case that domain library provides a pre-assembled pipeline with a `Header` that can limit the length. When users don't want to use it, it should be feasible to reset it back to no-op.

**Comment by mabu-github** [comment]
Do you have a code example for that case? It might be worth investigating a bit into the context for a better understanding. 

Generally one wants to avoid opening up the possibility to construct an instance in a way that leads to undefined behavior. Maybe there are options like differentiating between pipelines that do not have or require a length and ones that do. Then correctness could be ensured already by typing.

**Comment by ejguan** [comment]
Here is a similar example of `shuffle`.
https://github.com/pytorch/text/blob/2978507775283859e0519bbbd8e25527428d9a12/torchtext/datasets/ag_news.py#L41-L81

The pipeline provided from torchtext has shuffle disabled. Users can specify if they want to shuffle by giving an adaptor to `DataLoader2`.

```py
from torchdata.dataloader2 import DataLoader2
from torchdata.dataloader2.adapter import Shuffle

# Shuffle turned off by default
dl = DataLoader2(datapipe)

# Shuffle turned on
dl = DataLoader2(datapipe, datapipe_adapter_fn =Shuffle(True))

```
I would expect users having the similar way to specify `Header` as well
